R/attributes.R

```diff
@@ -530,34 +530,36 @@ vertex.attributes <- function(graph, index = V(graph)) {
   res
 }
 
+set_value_at <- function(value, idx, length_out) {
+  out <- value[NULL]
+  length(out) <- length_out
+  out[idx] <- value
+  unname(out)
+}
+
 #' @export
 "vertex.attributes<-" <- function(graph, index = V(graph), value) {
   ensure_igraph(graph)
 
   assert_named_list(value)
 
-  if (any(sapply(value, length) != length(index))) {
+  if (!all(lengths(value) == length(index))) {
     stop("Invalid attribute value length, must match number of vertices")
   }
 
   if (!missing(index)) {
     index <- as_igraph_vs(graph, index)
 
-    if (any(duplicated(index)) || any(is.na(index))) {
+    if (anyDuplicated(index) || anyNA(index)) {
       stop("Invalid vertices in index")
     }
   }
 
-  if (!missing(index) &&
-    (length(index) != vcount(graph) || any(index != V(graph)))) {
-    vs <- V(graph)
-    for (i in seq_along(value)) {
-      tmp <- value[[i]]
-      length(tmp) <- 0
-      length(tmp) <- length(vs)
-      tmp[index] <- value[[i]]
-      value[[i]] <- tmp
-    }
+  index_is_natural_sequence <- (length(index) == vcount(graph) && all(index == V(graph)))
```

Comment:
```suggestion
  index_is_natural_sequence <- (length(index) == vcount(graph) && identical(index, seq_len(vcount(graph))))
```

Reply:
Let's do `seq(1, vcount(graph))` instead.

Reply:
@krlmlr, @maelle,

This will not work if `vcount(graph) == 0`.
seq(1,0) returns `[1] 1 0`.